example/kokkos_pi/kokkos_pi.cpp

```diff
@@ -0,0 +1,299 @@
+
+// simple giga-ops calculator
+double calcGops(const int &nops, const double &t, const int &n_block3,
+                const int &n_mesh3, const int &n_iter) {
+  return (static_cast<Real>(nops*n_iter) / t / 1.0e9 * static_cast<Real>(n_block3) *
+          static_cast<Real>(n_mesh3));
+};
+
+// Test wrapper to run a function multiple times
+template <typename PerfFunc>
+double kernel_timer_wrapper(const int n_burn, const int n_perf, PerfFunc perf_func) {
+
+  // Initialize the timer and test
+  Kokkos::Timer timer;
+
+  for (int i_run = 0; i_run < n_burn + n_perf; i_run++) {
+
+    if (i_run == n_burn) {
+      // Burn in time is over, start timing
+      Kokkos::fence();
+      timer.reset();
+    }
+
+    // Run the function timing performance
+    perf_func();
+  }
+
+  // Time it
+  Kokkos::fence();
+  double perf_time = timer.seconds();
+
+  return perf_time;
+}
+
+void usage(std::string program) {
+  std::cout << std::endl
+            << "    Usage: " << program << " n_block n_mesh n_iter" << std::endl
+            << std::endl
+            << "             n_block = size of each mesh block on each axis" << std::endl
+            << "              n_mesh = number mesh blocks alogn each axis" << std::endl
+            << "              n_iter = number of iterations to time" << std::endl
+            << "            [Radius] = Optional: Radius of sphere" << std::endl
+            << "                                 Defaults to 1.0" << std::endl
+            << std::endl;
+}
+
+int main(int argc, char *argv[]) {
+  Kokkos::initialize(argc, argv);
+  do {
+
+    // ensure we have correct number of arguments
+    if (!(argc == 4 || argc == 5)) {
+      std::cout << "argc=" << argc << std::endl;
+      usage(argv[0]);
+      break;
+    }
+
+    std::size_t pos;
+    Real radius = 1.0;
+
+    // Read command line input
+    const int n_block = std::stoi(argv[1], &pos);
+    const int n_mesh = std::stoi(argv[2], &pos);
+    const int n_iter = std::stoi(argv[3], &pos);
+    if (argc == 5) {
+      radius = static_cast<Real>(std::stod(argv[4], &pos));
+    }
+
+    // Setup auxilliary variables
+    const int n_block2 = n_block * n_block;
+    const int n_block3 = n_block * n_block * n_block;
+    const int n_mesh3 = n_mesh * n_mesh * n_mesh;
+    const double radius2 = radius * radius;
+    const double radius3 = radius * radius * radius;
+    const Real dxyzCell = radius / static_cast<Real>(n_mesh * n_block);
+    const Real dVol = radius3 / static_cast<Real>(n_mesh3 * n_block3);
+
+    // allocate space for origin coordinates and create host mirror view
+    View2D xyz("xyzBlocks", 3, n_mesh3);
+    auto h_xyz = Kokkos::create_mirror_view(xyz);
+
+    // *** Kludge warning ***
+    // Since our mesh is not GPU friendly we set up a hacked up
+    // collection of mesh blocks.  The hope is that when our mesh is
+    // up to par we will replace this code with the mesh
+    // infrastructure.
```

Comment:
presumably once the mesh is up to par, we can just add your timing stuff directly into `calculate_pi` no?

Reply:
Good point.